File: clojure/bob/src/bob.clj
(ns bob)

(defn question?
  "Take a string, if it is a question return true, otherwise false"
  [string]
  (= \? (last (clojure.string/trim string))))

(defn phrase?
  "Take a string, return if it is a valid phrase"
  [string]
  (seq (filter #(Character/isLetter %) string)))

(defn yell?
  "Take a string, if it is an yell return true, otherwise false"
  [string]
  (if-let [validated-string (phrase? string)]
    (every? #(Character/isUpperCase %) validated-string)
    false))

(defn response-for
  "Exercism - Bob"
  [string]
  (cond
    (clojure.string/blank? string) "Fine. Be that way!"
    (and (question? string) (yell? string)) "Calm down, I know what I'm doing!"
    (yell? string) "Whoa, chill out!"
    (question? string) "Sure."
    :else "Whatever."))

File: elixir/bob/lib/bob.ex
defmodule Bob do
  def hey(input) do
    cond do
      nothing?(input) -> "Fine. Be that way!"
      question?(input) and yell?(input) -> "Calm down, I know what I'm doing!"
      question?(input) -> "Sure."
      yell?(input) -> "Whoa, chill out!"
      true -> "Whatever."
    end
  end

  defp question?(string) do
    string
    |> String.trim()
    |> String.ends_with?("?")
  end

  defp yell?(string) do
    string == String.upcase(string) and String.downcase(string) != String.upcase(string)
  end

  defp nothing?(input) do
    String.trim(input) == ""
  end
end
